src/app/ChunkedWriteCallback.h

```diff
@@ -0,0 +1,51 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/WriteClient.h>
+
+namespace chip {
+namespace app {
+
+/*
+ * This is an adapter that intercepts calls that deliver status codes from the WriteClient,
+ * selectively "merge"s the status code of the same attribute write for chunked list.
+ * - If the whole list was successfully written, callback->OnResponse will be called with success.
+ * - If any element in the list was not successfully written, callback->OnResponse will be called with the first error received.
+ * The merge logic is based on the list chunking behavior of WriteClient.
+ */
+class ChunkedWriteCallback : public WriteClient::Callback
+{
+public:
+    ChunkedWriteCallback(WriteClient::Callback * apCallback) : callback(apCallback) {}
+
+    void OnResponse(const app::WriteClient * apWriteClient, const app::ConcreteAttributePath & aPath,
```

Comment:
With chunking, the signature of the `OnResponse` must switch to `ConcreteDataAttributePath`, not `ConcreteAttributePath`. This is because in the `WriteResponse`, you'll get back status codes for each attribute path, which can target list elements.

The Adapter however will 'convert' those to effectively `ConcreteAttributePath` (but without altering the signature since it is an adapter). Upstream users of `ChunkedWriteCallback` can validate that (checkout `TypedReadAttributeCallback::OnAttributeData` for an example.

Reply:
Fixed